File: 0000xx/000097/java/Solution.java
import java.lang.AssertionError;
import java.util.*;
import java.util.regex.*;

// [WKZLRYSMSG]

class Solution {

    static class U implements Comparable{
        int done1;int done2;
        U(int done1,int done2){this.done1=done1;this.done2=done2;}
        public int compareTo(Object o){
            U u=(U)o;
            if(u.done1+u.done2!=this.done1+this.done2)return (this.done1+this.done2)-(u.done1+u.done2);
            if(this.done1!=u.done1)return this.done1-u.done1;
            if(this.done2!=u.done2)return this.done2-u.done2;
            return 0;
        }
        public boolean equals(Object o){
            return this.compareTo(o)==0;
        }
        public int hashCode(){
            return Objects.hash(done1,done2);
        }
    }

    public boolean isInterleave(String s1, String s2, String s3) {
        char[] c1Ary = s1.toCharArray();
        char[] c2Ary = s2.toCharArray();
        char[] c3Ary = s3.toCharArray();
        
        if(c1Ary.length+c2Ary.length!=c3Ary.length)return false;
        
        TreeSet<U> pQueue = new TreeSet<>();
        HashSet<U> uDoneSet = new HashSet<>();
        
        U u0=new U(0,0);
        pQueue.add(u0);uDoneSet.add(u0);
        while(!pQueue.isEmpty()){
            U u=pQueue.last();pQueue.remove(u);
            
            // go 1
            int done1 = u.done1;
            int done3 = u.done1+u.done2;
            while(true){
                if(done1>=c1Ary.length)break;
                if(c1Ary[done1]!=c1Ary[u.done1])break;
                if(c1Ary[done1]!=c3Ary[done3])break;
                ++done1;++done3;
            }
            if(done3>=c3Ary.length)return true;
            U newU = new U(done1,u.done2);
            if(!uDoneSet.contains(newU)){
                pQueue.add(newU);
                uDoneSet.add(newU);
            }
            
            // go 2
            int done2 = u.done2;
            done3 = u.done1+u.done2;
            while(true){
                if(done2>=c2Ary.length)break;
                if(c2Ary[done2]!=c2Ary[u.done2])break;
                if(c2Ary[done2]!=c3Ary[done3])break;
                ++done2;++done3;
            }
            if(done3>=c3Ary.length)return true;
            newU = new U(u.done1,done2);
            if(!uDoneSet.contains(newU)){
                pQueue.add(newU);
                uDoneSet.add(newU);
            }
        }
        
        return false;
    }
}

File: 0000xx/000097/kotlin/Solution.kt
class Solution {
    fun isInterleave(s1: String, s2: String, s3: String): Boolean {
        if(s3.length!=s1.length+s2.length)return false
        
        val c1Ary = s1.toCharArray()
        val c2Ary = s2.toCharArray()
        val c3Ary = s3.toCharArray()
        
        var goodAry = BooleanArray(s2.length+1){false}
        var nextGoodAry = BooleanArray(s2.length+1){false}
        
        goodAry[0] = true
        for(i2 in c2Ary.indices){
            goodAry[i2+1] = goodAry[i2] && (c2Ary[i2] == c3Ary[i2])
        }
        //trace(goodAry)
        
        for(i1 in c1Ary.indices){
            nextGoodAry[0] = goodAry[0] && (c1Ary[i1] == c3Ary[i1])
            for(i2 in c2Ary.indices){
                val c3 = c3Ary[i1+i2+1]
                nextGoodAry[i2+1] = (nextGoodAry[i2] && (c2Ary[i2] == c3) ) || (goodAry[i2+1] && (c1Ary[i1] == c3) )
            }
            val tmpGoodAry = goodAry
            goodAry = nextGoodAry
            nextGoodAry = tmpGoodAry
            //trace(goodAry)
        }
        
        return goodAry[s2.length]
    }
    
    fun trace(goodAry:BooleanArray){
        for(b in goodAry){
            print(if(b){"1 "}else{"0 "})
        }
        println("")
    }
}
